File: kotlin/src/main/kotlin/com/bhengubv/circleai/spatial/Spatial.kt
// Spatial.kt
//
// Kotlin port of CircleAI.Spatial (Contracts.cs + InMemorySpatial.cs +
// NullImplementations.cs) — the C# reference is the EXACT spec. Geo tiles +
// place search, synthetic radar, synthetic sky tracking, and a JSON (GLTF)
// 3D-scene renderer.
//
// Fidelity notes:
//   * C# `record` -> Kotlin `data class`.
//   * C# `ReadOnlyMemory<byte>` -> `ByteArray`.
//   * C# `DateTimeOffset` -> `java.time.Instant`; C# `Random(int)` -> `java.util.Random(long)`.
//   * C# `ValueTask` async members -> `suspend fun`.
//   * The synthetic radar seeds a Random from the coordinates so tests are
//     deterministic; the sky tracker filters by a daily-rotation altitude check.
//   * SearchPlacesAsync filters by name substring (OrdinalIgnoreCase), orders by
//     name, takes topK.

package com.bhengubv.circleai.spatial

import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonPrimitive
import java.time.Instant
import java.time.ZoneOffset
import java.util.Random
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.abs
import kotlin.math.cos
import kotlin.math.sin

// =====================================================================
// Contracts (Contracts.cs)
// =====================================================================

/** A latitude/longitude pair. Mirrors C# `LatLon`. */
data class LatLon(val latitude: Double, val longitude: Double)

/** A single map tile. Mirrors C# `GeoTile`. */
data class GeoTile(val z: Int, val x: Int, val y: Int, val imageBytes: ByteArray, val mimeType: String)

/** A single radar return. Mirrors C# `RadarReturn`. */
data class RadarReturn(val position: LatLon, val dopplerKmh: Double, val intensityDbz: Double)

/** A radar reading over a centre + range. Mirrors C# `RadarReading`. */
data class RadarReading(val centre: LatLon, val rangeKm: Double, val returns: List<RadarReturn>)

/** A visible sky object. Mirrors C# `SkyObject`. */
data class SkyObject(val name: String, val azimuthDeg: Double, val altitudeDeg: Double, val magnitudeApparent: Double)

/** An encoded 3D scene. Mirrors C# `Scene3D`. */
data class Scene3D(val sceneId: String, val encoded: ByteArray, val format: String)

/** Map-tile source (deck.gl / cesium pattern). Mirrors C# `IGeoTileSource`. */
interface IGeoTileSource {
    val backendId: String
    suspend fun getTileAsync(z: Int, x: Int, y: Int): GeoTile
    suspend fun searchPlacesAsync(query: String, topK: Int = 5): List<LatLon>
}

/** Weather / surveillance radar (RADAR pattern). Mirrors C# `IRadarReadout`. */
interface IRadarReadout {
    val backendId: String
    suspend fun getCurrentReadingAsync(at: LatLon, rangeKm: Double = 50.0): RadarReading
}

/** Visible-sky tracking (skylight pattern). Mirrors C# `ISkyTracker`. */
interface ISkyTracker {
    val backendId: String
    suspend fun visibleAsync(at: LatLon, utc: Instant): List<SkyObject>
}

/** 3D-scene rendering hook (flame / anime pattern). Mirrors C# `I3DSceneRenderer`. */
interface I3DSceneRenderer {
    val backendId: String
    suspend fun renderAsync(sceneScript: String, format: String = "gltf"): Scene3D
}

// =====================================================================
// In-memory implementations (InMemorySpatial.cs)
// =====================================================================

/** In-memory geo tile source with registered places. Mirrors C# `InMemoryGeoTileSource`. */
class InMemoryGeoTileSource : IGeoTileSource {
    private val places = ConcurrentHashMap<String, LatLon>()

    init {
        register("Johannesburg", LatLon(-26.2041, 28.0473))
        register("Cape Town", LatLon(-33.9249, 18.4241))
        register("Pretoria", LatLon(-25.7479, 28.2293))
        register("Durban", LatLon(-29.8587, 31.0218))
        register("Lagos", LatLon(6.5244, 3.3792))
        register("Nairobi", LatLon(-1.2921, 36.8219))
        register("London", LatLon(51.5074, -0.1278))
        register("New York", LatLon(40.7128, -74.0060))
    }

    override val backendId: String get() = "in-memory"

    fun register(name: String, at: LatLon) {
        require(name.isNotBlank()) { "name required" }
        places[name] = at
    }

    override suspend fun getTileAsync(z: Int, x: Int, y: Int): GeoTile {
        if (z < 0 || x < 0 || y < 0) throw IndexOutOfBoundsException("z")
        // 1x1 transparent PNG.
        val pngBytes = byteArrayOf(
            0x89.toByte(), 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x00, 0x00, 0x00, 0x0D, 0x49, 0x48, 0x44, 0x52,
            0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1F, 0x15, 0xC4.toByte(),
            0x89.toByte(), 0x00, 0x00, 0x00, 0x0D, 0x49, 0x44, 0x41, 0x54, 0x78, 0x9C.toByte(), 0x63, 0x00, 0x01, 0x00, 0x00,
            0x05, 0x00, 0x01, 0x0D, 0x0A, 0x2D, 0xB4.toByte(), 0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4E, 0x44, 0xAE.toByte(),
            0x42, 0x60, 0x82.toByte(),
        )
        return GeoTile(z, x, y, pngBytes, "image/png")
    }

    override suspend fun searchPlacesAsync(query: String, topK: Int): List<LatLon> {
        if (topK <= 0) throw IndexOutOfBoundsException("topK")
        return places.entries
            .filter { it.key.contains(query, ignoreCase = true) }
            .sortedBy { it.key }
            .take(topK)
            .map { it.value }
    }
}

/** Deterministic synthetic radar. Mirrors C# `SyntheticRadarReadout`. */
class SyntheticRadarReadout : IRadarReadout {
    override val backendId: String get() = "synthetic"

    override suspend fun getCurrentReadingAsync(at: LatLon, rangeKm: Double): RadarReading {
        if (rangeKm <= 0) throw IndexOutOfBoundsException("rangeKm")

        val seed = (at.latitude * 1000).toLong() + (at.longitude * 1000).toLong() + (rangeKm * 10).toLong()
        val rng = Random((seed xor (seed shr 32)))
        val count = 3 + rng.nextInt(5)
        val rets = ArrayList<RadarReturn>(count)
        for (i in 0 until count) {
            val d = rng.nextDouble() * rangeKm * 0.9
            val ang = rng.nextDouble() * Math.PI * 2
            val lat = at.latitude + (cos(ang) * d) / 111.0
            val lon = at.longitude + (sin(ang) * d) / 111.0
            rets.add(RadarReturn(LatLon(lat, lon), rng.nextDouble() * 60 - 30, rng.nextDouble() * 60))
        }
        return RadarReading(at, rangeKm, rets)
    }
}

/** Deterministic synthetic sky tracker. Mirrors C# `SyntheticSkyTracker`. */
class SyntheticSkyTracker : ISkyTracker {
    override val backendId: String get() = "synthetic"

    override suspend fun visibleAsync(at: LatLon, utc: Instant): List<SkyObject> {
        val tod = utc.atOffset(ZoneOffset.UTC).toLocalTime()
        val hours = tod.toNanoOfDay() / 3_600_000_000_000.0
        val rot = hours * 15.0 // earth rotation degrees-per-hour
        val hits = ArrayList<SkyObject>(BASE_OBJECTS.size)
        for ((n, az, alt, mag) in BASE_OBJECTS) {
            val az2 = (az - rot + 360) % 360
            if (alt - abs(at.latitude) > 0) {
                hits.add(SkyObject(n, az2, alt, mag))
            }
        }
        return hits
    }

    private companion object {
        val BASE_OBJECTS = arrayOf(
            SkyBase("Sirius", 102.7, 35.0, -1.46),
            SkyBase("Polaris", 0.0, 51.5, 1.97),
            SkyBase("Vega", 88.0, 70.0, 0.03),
            SkyBase("Mars", 135.4, 22.0, 0.5),
            SkyBase("Jupiter", 180.5, 40.0, -2.0),
            SkyBase("Saturn", 210.0, 30.0, 0.4),
        )
    }

    private data class SkyBase(val name: String, val azimuth: Double, val altitude: Double, val mag: Double)
}

/** Minimal-GLTF JSON 3D-scene renderer. Mirrors C# `JsonScene3DRenderer`. */
class JsonScene3DRenderer : I3DSceneRenderer {
    override val backendId: String get() = "json"

    override suspend fun renderAsync(sceneScript: String, format: String): Scene3D {
        val fmt = format.ifBlank { "gltf" }
        val sceneId = UUID.randomUUID().toString().replace("-", "")
        val escapedScript = Json.encodeToString(JsonPrimitive.serializer(), JsonPrimitive(sceneScript))
        val json = "{\"asset\":{\"version\":\"2.0\",\"generator\":\"CircleAI.Spatial.JsonScene3DRenderer\"}," +
            "\"scenes\":[{\"nodes\":[]}],\"scene\":0,\"extras\":{\"script\":$escapedScript}}"
        return Scene3D(sceneId, json.toByteArray(Charsets.UTF_8), fmt)
    }
}

// =====================================================================
// Null implementations (NullImplementations.cs)
// =====================================================================

private const val SPATIAL_EMPTY_GUID = "00000000-0000-0000-0000-000000000000"

/** No-op [IGeoTileSource]. Mirrors C# `NullGeoTileSource`. */
class NullGeoTileSource private constructor() : IGeoTileSource {
    override val backendId: String get() = "null"
    override suspend fun getTileAsync(z: Int, x: Int, y: Int): GeoTile =
        GeoTile(z, x, y, ByteArray(0), "image/png")
    override suspend fun searchPlacesAsync(query: String, topK: Int): List<LatLon> = emptyList()

    companion object {
        val Instance = NullGeoTileSource()
    }
}

/** No-op [IRadarReadout]. Mirrors C# `NullRadarReadout`. */
class NullRadarReadout private constructor() : IRadarReadout {
    override val backendId: String get() = "null"
    override suspend fun getCurrentReadingAsync(at: LatLon, rangeKm: Double): RadarReading =
        RadarReading(at, rangeKm, emptyList())

    companion object {
        val Instance = NullRadarReadout()
    }
}

/** No-op [ISkyTracker]. Mirrors C# `NullSkyTracker`. */
class NullSkyTracker private constructor() : ISkyTracker {
    override val backendId: String get() = "null"
    override suspend fun visibleAsync(at: LatLon, utc: Instant): List<SkyObject> = emptyList()

    companion object {
        val Instance = NullSkyTracker()
    }
}

/** No-op [I3DSceneRenderer]. Mirrors C# `Null3DSceneRenderer`. */
class Null3DSceneRenderer private constructor() : I3DSceneRenderer {
    override val backendId: String get() = "null"
    override suspend fun renderAsync(sceneScript: String, format: String): Scene3D =
        Scene3D(SPATIAL_EMPTY_GUID, ByteArray(0), format)

    companion object {
        val Instance = Null3DSceneRenderer()
    }
}

File: swift/Sources/CircleAI/Spatial.swift
// Spatial.swift
//
// Port of src/CircleAI.Spatial/:
//   • Contracts.cs                 — LatLon, GeoTile, RadarReading,
//                                     RadarReturn, SkyObject, Scene3D;
//                                     IGeoTileSource, IRadarReadout,
//                                     ISkyTracker, I3DSceneRenderer
//   • InMemorySpatial.cs           — InMemoryGeoTileSource (1x1 PNG + registered
//                                     place search), SyntheticRadarReadout
//                                     (deterministic seeded pattern),
//                                     SyntheticSkyTracker (rotation filter),
//                                     JsonScene3DRenderer (minimal GLTF 2.0)
//   • NullImplementations.cs       — Null* backends
//
// Porting notes:
//   • `record` → `struct: Sendable`. `ReadOnlyMemory<byte>` → `[UInt8]`.
//     GeoTile / Scene3D hold bytes → Sendable + Equatable only.
//   • `System.Random(seed)` → `DotNetRandom(seed:)` (bit-identical .NET RNG,
//     already in the tree), with `next(_:_:)` for `rng.Next(0,5)`.
//   • The radar seed math replicates the C# Int64 arithmetic + XOR-fold to Int32
//     exactly, so the deterministic pattern matches.
//   • `Guid.NewGuid():n` → UUID hex. GLTF script embed uses
//     `JSONSerialization` to JSON-encode the script string (matches
//     `System.Text.Json.JsonSerializer.Serialize(string)`).
//   • Guards → `SpatialError`.

import Foundation

// MARK: - Records

/// A latitude/longitude pair (WGS-84 degrees).
public struct LatLon: Sendable, Equatable, Codable {
    public let latitude: Double
    public let longitude: Double
    public init(latitude: Double, longitude: Double) {
        self.latitude = latitude
        self.longitude = longitude
    }
}

/// A rendered map tile (z/x/y) plus its image bytes and MIME type.
public struct GeoTile: Sendable, Equatable {
    public let z: Int
    public let x: Int
    public let y: Int
    public let imageBytes: [UInt8]
    public let mimeType: String
    public init(z: Int, x: Int, y: Int, imageBytes: [UInt8], mimeType: String) {
        self.z = z
        self.x = x
        self.y = y
        self.imageBytes = imageBytes
        self.mimeType = mimeType
    }
}

/// A single radar return (position + doppler + intensity).
public struct RadarReturn: Sendable, Equatable, Codable {
    public let position: LatLon
    public let dopplerKmh: Double
    public let intensityDbz: Double
    public init(position: LatLon, dopplerKmh: Double, intensityDbz: Double) {
        self.position = position
        self.dopplerKmh = dopplerKmh
        self.intensityDbz = intensityDbz
    }
}

/// A radar sweep centred at a point with a range and a set of returns.
public struct RadarReading: Sendable, Equatable, Codable {
    public let centre: LatLon
    public let rangeKm: Double
    public let returns: [RadarReturn]
    public init(centre: LatLon, rangeKm: Double, returns: [RadarReturn]) {
        self.centre = centre
        self.rangeKm = rangeKm
        self.returns = returns
    }
}

/// A visible sky object (star / planet) with az/alt and apparent magnitude.
public struct SkyObject: Sendable, Equatable, Codable {
    public let name: String
    public let azimuthDeg: Double
    public let altitudeDeg: Double
    public let magnitudeApparent: Double
    public init(name: String, azimuthDeg: Double, altitudeDeg: Double, magnitudeApparent: Double) {
        self.name = name
        self.azimuthDeg = azimuthDeg
        self.altitudeDeg = altitudeDeg
        self.magnitudeApparent = magnitudeApparent
    }
}

/// A rendered 3D scene (encoded bytes + format tag).
public struct Scene3D: Sendable, Equatable {
    public let sceneId: String
    public let encoded: [UInt8]
    public let format: String
    public init(sceneId: String, encoded: [UInt8], format: String) {
        self.sceneId = sceneId
        self.encoded = encoded
        self.format = format
    }
}

// MARK: - Errors

public enum SpatialError: Error, Equatable, CustomStringConvertible {
    case nameRequired
    case coordinateOutOfRange
    case topKOutOfRange
    case rangeKmOutOfRange

    public var description: String {
        switch self {
        case .nameRequired: return "name required"
        case .coordinateOutOfRange: return "z"
        case .topKOutOfRange: return "topK"
        case .rangeKmOutOfRange: return "rangeKm"
        }
    }
}

// MARK: - Contracts

public protocol IGeoTileSource: Sendable {
    var backendId: String { get }
    func getTile(z: Int, x: Int, y: Int) async throws -> GeoTile
    func searchPlaces(query: String, topK: Int) async throws -> [LatLon]
}

public protocol IRadarReadout: Sendable {
    var backendId: String { get }
    func getCurrentReading(at: LatLon, rangeKm: Double) async throws -> RadarReading
}

public protocol ISkyTracker: Sendable {
    var backendId: String { get }
    func visible(at: LatLon, utc: Date) async throws -> [SkyObject]
}

public protocol I3DSceneRenderer: Sendable {
    var backendId: String { get }
    func render(sceneScript: String, format: String) async throws -> Scene3D
}

// MARK: - In-memory geo-tile source

/// Deterministic geo-tile source: returns a 1x1 transparent PNG for every tile
/// and searches a registered gazetteer of place names (case-insensitive
/// substring, ordered ascending by name).
public final class InMemoryGeoTileSource: IGeoTileSource, @unchecked Sendable {
    private let lock = NSLock()
    private var places: [String: LatLon] = [:]

    public init() {
        register("Johannesburg", LatLon(latitude: -26.2041, longitude: 28.0473))
        register("Cape Town", LatLon(latitude: -33.9249, longitude: 18.4241))
        register("Pretoria", LatLon(latitude: -25.7479, longitude: 28.2293))
        register("Durban", LatLon(latitude: -29.8587, longitude: 31.0218))
        register("Lagos", LatLon(latitude: 6.5244, longitude: 3.3792))
        register("Nairobi", LatLon(latitude: -1.2921, longitude: 36.8219))
        register("London", LatLon(latitude: 51.5074, longitude: -0.1278))
        register("New York", LatLon(latitude: 40.7128, longitude: -74.0060))
    }

    public var backendId: String { "in-memory" }

    public func register(_ name: String, _ at: LatLon) {
        lock.lock(); defer { lock.unlock() }
        places[name] = at
    }

    public func getTile(z: Int, x: Int, y: Int) async throws -> GeoTile {
        if z < 0 || x < 0 || y < 0 { throw SpatialError.coordinateOutOfRange }
        // 1x1 transparent PNG.
        let png: [UInt8] = [
            0x89, 0x50, 0x4E, 0x47, 0x0D, 0x0A, 0x1A, 0x0A, 0x00, 0x00, 0x00, 0x0D, 0x49, 0x48, 0x44, 0x52,
            0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x01, 0x08, 0x06, 0x00, 0x00, 0x00, 0x1F, 0x15, 0xC4,
            0x89, 0x00, 0x00, 0x00, 0x0D, 0x49, 0x44, 0x41, 0x54, 0x78, 0x9C, 0x63, 0x00, 0x01, 0x00, 0x00,
            0x05, 0x00, 0x01, 0x0D, 0x0A, 0x2D, 0xB4, 0x00, 0x00, 0x00, 0x00, 0x49, 0x45, 0x4E, 0x44, 0xAE,
            0x42, 0x60, 0x82,
        ]
        return GeoTile(z: z, x: x, y: y, imageBytes: png, mimeType: "image/png")
    }

    public func searchPlaces(query: String, topK: Int = 5) async throws -> [LatLon] {
        if topK <= 0 { throw SpatialError.topKOutOfRange }
        lock.lock(); defer { lock.unlock() }
        let hits = places
            .filter { $0.key.range(of: query, options: .caseInsensitive) != nil }
            .sorted { $0.key < $1.key }
            .prefix(topK)
            .map { $0.value }
        return Array(hits)
    }
}

// MARK: - Synthetic radar

/// Deterministic radar readout — the return pattern is fully determined by the
/// coordinates and range, seeded via the byte-identical .NET RNG so results are
/// reproducible.
public struct SyntheticRadarReadout: IRadarReadout {
    public init() {}
    public var backendId: String { "synthetic" }

    public func getCurrentReading(at: LatLon, rangeKm: Double = 50) async throws -> RadarReading {
        if rangeKm <= 0 { throw SpatialError.rangeKmOutOfRange }

        // Deterministic seed — replicates the C# Int64 arithmetic + XOR-fold.
        let seed: Int64 = Int64(at.latitude * 1000) + Int64(at.longitude * 1000) + Int64(rangeKm * 10)
        let seed32 = Int32(truncatingIfNeeded: seed ^ (seed >> 32))
        var rng = DotNetRandom(seed: seed32)

        let count = 3 + rng.next(0, 5)
        var returns: [RadarReturn] = []
        returns.reserveCapacity(count)
        for _ in 0..<count {
            let d = rng.nextDouble() * rangeKm * 0.9
            let ang = rng.nextDouble() * Double.pi * 2
            let lat = at.latitude + (cos(ang) * d) / 111.0
            let lon = at.longitude + (sin(ang) * d) / 111.0
            returns.append(RadarReturn(
                position: LatLon(latitude: lat, longitude: lon),
                dopplerKmh: rng.nextDouble() * 60 - 30,
                intensityDbz: rng.nextDouble() * 60))
        }
        return RadarReading(centre: at, rangeKm: rangeKm, returns: returns)
    }
}

// MARK: - Synthetic sky tracker

/// Deterministic sky tracker — a fixed catalogue of bright objects rotated by
/// time-of-day and filtered by a crude altitude/latitude visibility rule.
public struct SyntheticSkyTracker: ISkyTracker {
    private static let baseObjects: [(name: String, azimuth: Double, altitude: Double, mag: Double)] = [
        ("Sirius", 102.7, 35.0, -1.46),
        ("Polaris", 0.0, 51.5, 1.97),
        ("Vega", 88.0, 70.0, 0.03),
        ("Mars", 135.4, 22.0, 0.5),
        ("Jupiter", 180.5, 40.0, -2.0),
        ("Saturn", 210.0, 30.0, 0.4),
    ]

    public init() {}
    public var backendId: String { "synthetic" }

    public func visible(at: LatLon, utc: Date) async throws -> [SkyObject] {
        // hours = fractional hours since UTC midnight of `utc`.
        var cal = Calendar(identifier: .gregorian)
        cal.timeZone = TimeZone(identifier: "UTC")!
        let comps = cal.dateComponents([.hour, .minute, .second, .nanosecond], from: utc)
        let hH  = Double(comps.hour ?? 0)
        let hM  = Double(comps.minute ?? 0) / 60.0
        let hS  = Double(comps.second ?? 0) / 3600.0
        let hNs = Double(comps.nanosecond ?? 0) / 3_600_000_000_000.0
        let hours = hH + hM + hS + hNs
        let rot = hours * 15.0 // earth rotation degrees-per-hour

        var hits: [SkyObject] = []
        for (n, az, alt, mag) in SyntheticSkyTracker.baseObjects {
            let az2 = (az - rot + 360).truncatingRemainder(dividingBy: 360)
            if alt - abs(at.latitude) > 0 {
                hits.append(SkyObject(name: n, azimuthDeg: az2, altitudeDeg: alt, magnitudeApparent: mag))
            }
        }
        return hits
    }
}

// MARK: - JSON 3D-scene renderer

/// Wraps a scene script into a minimal valid GLTF 2.0 JSON document (the script
/// is stored as a JSON-escaped `extras.script` blob).
public struct JsonScene3DRenderer: I3DSceneRenderer {
    public init() {}
    public var backendId: String { "json" }

    public func render(sceneScript: String, format: String = "gltf") async throws -> Scene3D {
        let fmt = format.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty ? "gltf" : format
        let sceneId = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()

        // JSON-encode the script string exactly as System.Text.Json would.
        let scriptData = (try? JSONSerialization.data(withJSONObject: sceneScript, options: [.fragmentsAllowed])) ?? Data("\"\"".utf8)
        let scriptLiteral = String(decoding: scriptData, as: UTF8.self)

        let json = "{\"asset\":{\"version\":\"2.0\",\"generator\":\"CircleAI.Spatial.JsonScene3DRenderer\"},\"scenes\":[{\"nodes\":[]}],\"scene\":0,\"extras\":{\"script\":\(scriptLiteral)}}"
        return Scene3D(sceneId: sceneId, encoded: Array(json.utf8), format: fmt)
    }
}

// MARK: - Null backends

public struct NullGeoTileSource: IGeoTileSource {
    public static let instance = NullGeoTileSource()
    public init() {}
    public var backendId: String { "null" }
    public func getTile(z: Int, x: Int, y: Int) async throws -> GeoTile {
        GeoTile(z: z, x: x, y: y, imageBytes: [], mimeType: "image/png")
    }
    public func searchPlaces(query: String, topK: Int = 5) async throws -> [LatLon] { [] }
}

public struct NullRadarReadout: IRadarReadout {
    public static let instance = NullRadarReadout()
    public init() {}
    public var backendId: String { "null" }
    public func getCurrentReading(at: LatLon, rangeKm: Double = 50) async throws -> RadarReading {
        RadarReading(centre: at, rangeKm: rangeKm, returns: [])
    }
}

public struct NullSkyTracker: ISkyTracker {
    public static let instance = NullSkyTracker()
    public init() {}
    public var backendId: String { "null" }
    public func visible(at: LatLon, utc: Date) async throws -> [SkyObject] { [] }
}

public struct Null3DSceneRenderer: I3DSceneRenderer {
    public static let instance = Null3DSceneRenderer()
    public init() {}
    public var backendId: String { "null" }
    public func render(sceneScript: String, format: String = "gltf") async throws -> Scene3D {
        Scene3D(sceneId: "00000000-0000-0000-0000-000000000000", encoded: [], format: format)
    }
}
